core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1963 @@
+        self.entries.flush();
+        self.kv_pairs.flush();
+    }
+}
+
+impl<K, V> Encode for BTreeMap<K, V> {
+    fn encode_to<W: scale::Output>(&self, dest: &mut W) {
+        self.header.encode_to(dest);
+        self.entries.encode_to(dest);
+        self.kv_pairs.encode_to(dest);
+    }
+}
+
+impl<K, V> Decode for BTreeMap<K, V> {
+    fn decode<I: scale::Input>(input: &mut I) -> Result<Self, scale::Error> {
+        let header = storage::Value::decode(input)?;
+        let entries = SyncChunk::decode(input)?;
+        let kv_pairs = SyncChunk::decode(input)?;
+        Ok(Self {
+            header,
+            entries,
+            kv_pairs,
+        })
+    }
+}
+
+impl<K, V> AllocateUsing for BTreeMap<K, V> {
+    #[inline]
+    unsafe fn allocate_using<A>(alloc: &mut A) -> Self
+    where
+        A: Allocate,
+    {
+        Self {
+            header: storage::Value::allocate_using(alloc),
+            entries: SyncChunk::allocate_using(alloc),
+            kv_pairs: SyncChunk::allocate_using(alloc),
+        }
+    }
+}
+
+impl<K, V> BTreeMap<K, V> {
+    /// Returns the index of the root node.
+    pub(super) fn root(&self) -> &Option<NodeHandle> {
+        &self.header.root
+    }
+}
+
+impl<K, V> BTreeMap<K, V>
+where
+    K: Codec + Ord,
+    V: Codec,
+{
+    /// Returns the `HandleType` of `handle`. Either `Leaf` or `Internal`.
+    /// A internal node always has children, whereas a leaf doesn't.
+    pub(super) fn get_handle_type(&self, handle: NodeHandle) -> HandleType {
+        let children = self
+            .get_node(handle)
+            .expect("node must exist")
+            .edges_count();
+        if children == 0 {
+            Leaf
+        } else {
+            Internal
+        }
+    }
+
+    /// Returns the number of nodes which this tree consists of.
+    pub(super) fn node_count(&self) -> u32 {
+        self.header.node_count
+    }
+
+    /// Descends to the child node pointed to by this `handle`, if one is available.
+    pub(super) fn descend(&self, handle: KVHandle) -> Option<NodeHandle> {
+        let node = self
+            .get_node(handle.node())
+            .expect("node to descend from must exist");
+        node.edge(handle.idx())
+    }
+
+    /// Returns a reference to the node behind `handle`, if existent.
+    ///
+    /// *Note*
+    /// In practice this should never return `None` (except for tests which
+    /// explicitly test this). The calling side should apply an `.expect()`
+    /// to the return value and proof why this can never happen.
+    pub(super) fn get_node(&self, handle: NodeHandle) -> Option<&Node<K, V>> {
+        let entry = self.entries.get(handle.node())?;
+        match entry {
+            InternalEntry::Occupied(occupied) => Some(occupied),
+            InternalEntry::Vacant(_) => None,
+        }
+    }
+
+    /// Returns the keys stored in this node.
+    pub(super) fn keys_in_node(&self, node: NodeHandle) -> [Option<&K>; CAPACITY] {
```

Comment:
Wouldn't it be better if we instead return `&[K]` here?
Or is it possible that the resulting array has holes in it? E.g. `[Some, Some, None, Some]`

Reply:
I would leave it at `[Option<&K>]` since this method is used by the node search stuff, which e.g. knows that search can be stopped once a `None` element is reached.